File: jarvis/ui/controller.py
"""AppController: maps UI actions to JarvisService calls and posts result cards to the feed.

Pure (no Flet), so every action is unit-tested with a faked service; `ui/app.py` wires Flet widgets
to these methods. The controller holds NO business logic of its own - it only calls the facade and
shapes the result into a feed card. Shortcut-button actions (Slice 3) extend this same class.
"""

from __future__ import annotations

from jarvis.redact import redact
from jarvis.service import JarvisService
from jarvis.ui.feed import Card, Feed

# The "Markets/News" shortcut is a preset question through the same grounded ask path.
_MARKETS_QUERY = "What's happening in markets and tech news today?"


class AppController:
    def __init__(self, service: JarvisService, feed: Feed) -> None:
        self._service = service
        self._feed = feed

    def show_briefing(self) -> None:
        try:
            self._feed.post_card(Card("Daily briefing", self._service.briefing(), "briefing"))
        except Exception as exc:
            self._post_error(exc)

    def ask(self, text: str) -> None:
        text = text.strip()
        if not text:
            return
        self._feed.post_card(Card("You", text, "chat"))  # echo the question
        try:
            result = self._service.ask(text)
        except Exception as exc:
            self._post_error(exc)
            return
        title = "Jarvis (cached)" if result.cached else "Jarvis"
        self._feed.post_card(Card(title, result.text, "answer" if result.grounded else "chat"))

    # --- shortcut-button actions (Slice 3) -------------------------------------------------

    def show_agenda(self) -> None:
        try:
            result = self._service.agenda()
        except Exception as exc:
            self._post_error(exc)
            return
        if not result.connected:
            body = "Not connected. Run: python -m jarvis calendar-auth"
        elif not result.events:
            body = "No events today."
        else:
            body = "\n".join(_event_line(event) for event in result.events)
        self._feed.post_card(Card("Today's calendar", body, "agenda"))

    def ask_markets_news(self) -> None:
        self.ask(_MARKETS_QUERY)  # already guarded via ask()

    def add_goal(self, text: str) -> None:
        text = text.strip()
        if not text:
            return
        try:
            goal = self._service.add_goal(text)
        except Exception as exc:
            self._post_error(exc)
            return
        self._feed.post_card(Card("Goal added", f"#{goal.id}  {goal.description}", "goal"))

    def _post_error(self, exc: Exception) -> None:
        # A backend failure (e.g. Ollama down) posts an error card instead of crashing the GUI, so
        # the feed keeps receiving. Redacted, since exception text reaches a user surface.
        self._feed.post_card(Card("Error", redact(str(exc)), "error"))


def _event_line(event) -> str:
    when = "all day" if event.all_day else f"{event.start:%H:%M}-{event.end:%H:%M}"
    location = f" @ {event.location}" if event.location else ""
    return f"- {when} {event.summary}{location}"

File: frontend/src/controller.ts
// The card-shaping controller: the TS mirror of jarvis/ui/controller.py (AppController) + design
// spec §8. Each front-end owns its own presentation; this one calls the API and shapes the exact
// same cards the Flet GUI does (titles, the cached suffix, agenda/finance/goal line formats, the
// suggestion WhyStrip, redacted error cards). NO capability logic lives here.
import { api, ApiError } from "./api";
import type { CalendarEvent, CardData, Suggestion } from "./types";

export const MARKETS_QUERY = "What's happening in markets and tech news today?";
const MONTHLY_SPEND_QUERY = "How much have I spent this month?";

export type Post = (card: CardData) => void;

/** "HH:MM" in the event's own clock (slice the ISO time, no tz reinterpretation). */
function wallTime(iso: string): string {
  const m = /T(\d{2}:\d{2})/.exec(iso);
  return m ? m[1] : iso;
}

function eventLine(e: CalendarEvent): string {
  const when = e.all_day ? "all day" : `${wallTime(e.start)}-${wallTime(e.end)}`;
  const location = e.location ? ` @ ${e.location}` : "";
  return `- ${when} ${e.summary}${location}`;
}

export function makeController(post: Post) {
  const postError = (exc: unknown) => {
    // Backend failures arrive already redacted; transport errors carry a generic message.
    const body = exc instanceof ApiError ? exc.message : "Something went wrong.";
    post({ title: "Error", body, kind: "error" });
  };

  async function ask(text: string): Promise<void> {
    const t = text.trim();
    if (!t) return; // empty input posts nothing (mirrors the early return)
    post({ title: "You", body: t, kind: "chat" }); // echo the question
    try {
      const r = await api.ask(t);
      const title = r.cached ? "Jarvis (cached)" : "Jarvis";
      post({ title, body: r.text, kind: r.grounded ? "answer" : "chat" });
    } catch (exc) {
      postError(exc);
    }
  }

  async function showBriefing(): Promise<void> {
    try {
      const r = await api.briefing();
      post({ title: "Daily briefing", body: r.text, kind: "briefing" });
    } catch (exc) {
      postError(exc);
    }
  }

  async function showAgenda(): Promise<void> {
    try {
      const r = await api.agenda();
      let body: string;
      if (!r.connected) body = "Not connected. Run: python -m jarvis calendar-auth";
      else if (r.events.length === 0) body = "No events today.";
      else body = r.events.map(eventLine).join("\n");
      post({ title: "Today's calendar", body, kind: "agenda" });
    } catch (exc) {
      postError(exc);
    }
  }

  const askMarketsNews = () => ask(MARKETS_QUERY);

  async function showFinance(): Promise<void> {
    try {
      const r = await api.financeAsk(MONTHLY_SPEND_QUERY);
      post({ title: "Finance", body: r.text, kind: "finance" });
    } catch (exc) {
      postError(exc);
    }
  }

  async function addGoal(text: string): Promise<void> {
    const t = text.trim();
    if (!t) return;
    try {
      const goal = await api.addGoal(t);
      post({ title: "Goal added", body: `#${goal.id}  ${goal.description}`, kind: "goal" });
    } catch (exc) {
      postError(exc);
    }
  }

  async function showSuggestions(): Promise<Suggestion[]> {
    try {
      const { suggestions } = await api.suggestions();
      if (suggestions.length === 0) {
        post({ title: "Jarvis", body: "Nothing worth surfacing right now.", kind: "suggestion" });
        return [];
      }
      for (const s of suggestions) {
        post({ title: "Suggestion", body: s.content, kind: "suggestion", why: s.why });
      }
      return suggestions;
    } catch (exc) {
      postError(exc);
      return [];
    }
  }

  // The goal-driven PULL feed, surfaced INTO the readable feed on demand: one card per active goal
  // with relevant info, reusing the 'suggestion' kind so its WhyStrip explains the relevance.
  async function showGoalFeed(): Promise<void> {
    try {
      const { feed } = await api.goalFeed();
      const withItems = feed.filter((f) => f.items.length > 0);
      if (withItems.length === 0) {
        post({
          title: "Goal feed",
          body: "No goal-relevant info surfaced right now.",
          kind: "suggestion",
        });
        return;
      }
      for (const f of withItems) {
        const body = f.items
          .map((it) => {
            const link = it.url ? ` ([${it.source}](${it.url}))` : ` _(${it.source})_`;
            return `- **${it.title}** — ${it.detail}${link}`;
          })
          .join("\n");
        post({
          title: `Goal #${f.goal_id}: ${f.goal}`,
          body,
          kind: "suggestion",
          why: `Surfaced for goal #${f.goal_id}: ${f.goal}`,
        });
      }
    } catch (exc) {
      postError(exc);
    }
  }

  // The off-loopback write guard (jarvis/api/app.py) returns 503 with a "set JARVIS_API_TOKEN"
  // message. Surface it as a clear, actionable card - it is config, not a crash.
  const isWriteGuard = (exc: unknown) =>
    exc instanceof ApiError && exc.status === 503 && /JARVIS_API_TOKEN/.test(exc.message);

  const postFsError = (exc: unknown) => {
    if (isWriteGuard(exc)) {
      post({
        title: "File writes locked",
        body:
          (exc as ApiError).message +
          "\n\nThis cockpit is reachable off-localhost without a token. Set JARVIS_API_TOKEN on " +
          "the server and open it once via ?token=… to enable writes.",
        kind: "error",
      });
      return;
    }
    postError(exc); // bad path, permission denied, etc. - already redacted by the backend
  };

  async function createFile(path: string, content: string): Promise<void> {
    const p = path.trim();
    if (!p) return;
    try {
      const r = await api.createFile(p, content);
      const size = r.bytes != null ? `  (${r.bytes} B)` : "";
      post({ title: "File created", body: `\`${r.path}\`${size}`, kind: "answer" });
    } catch (exc) {
      postFsError(exc);
    }
  }

  async function createFolder(path: string): Promise<void> {
    const p = path.trim();
    if (!p) return;
    try {
      const r = await api.createFolder(p);
      const verb = r.created ? "Folder created" : "Folder already exists";
      post({ title: verb, body: `\`${r.path}\``, kind: "answer" });
    } catch (exc) {
      postFsError(exc);
    }
  }

  return {
    ask,
    showBriefing,
    showAgenda,
    askMarketsNews,
    showFinance,
    addGoal,
    showSuggestions,
    showGoalFeed,
    createFile,
    createFolder,
  };
}

export type Controller = ReturnType<typeof makeController>;
